# Functions Assignment

1. What is the difference between a function and a method in Python?
->

In Python programming, the terms "function" and "method" are often used interchangeably, but they refer to distinct concepts. Understanding the differences between them is fundamental to grasping how Python operates. This distinction plays a significant role in object-oriented programming (OOP) and procedural programming. In this assignment, we will explore and contrast functions and methods in Python, focusing on their definitions, usage, and key differences.

Definition of a Function

A function in Python is a block of code that is designed to perform a specific task. It can take inputs, process them, and return an output. Functions are defined using the def keyword followed by the function name and parameters. Functions are independent entities and can be called from anywhere within the program.

Syntax of a Function:

In [1]:
def function_name(parameter1, parameter2):
    # Function body
    return result


#Example:

In [2]:
def add_numbers(a, b):
    return a + b


#Key Differences Between a Function and a Method
#Definition:

 * A function is a block of reusable code that is not tied to any object.

 * A method is a function that is associated with an object and is defined within a class.

Invocation:

 * A function can be called independently, without needing an object or class.

 * A method must be called on an instance of a class (object).

#Parameters:

 * A function accepts parameters explicitly when called.

 * A method implicitly receives the instance of the class (usually named self) as the first parameter.

#Usage Context:

 * A function is used in procedural programming and can exist outside of classes.

 * A method is used in object-oriented programming (OOP) and is always part of a class.

#Example to Illustrate the Difference

In [3]:
# Defining a function
def multiply(x, y):
    return x * y

# Defining a class with a method
class MathOperations:
    def multiply(self, x, y):
        return x * y

# Calling the function
result_function = multiply(3, 4)

# Creating an object of MathOperations class
math_obj = MathOperations()

# Calling the method on the object
result_method = math_obj.multiply(3, 4)

print("Result from function:", result_function)
print("Result from method:", result_method)


Result from function: 12
Result from method: 12


#2. Explain the concept of function arguments and parameters in Python.
->

Functions are a core part of Python programming. They allow for modularity, code reuse, and better organization. An essential aspect of functions is the use of parameters and arguments, which help in passing data to functions. In this assignment, we will explore the concepts of function parameters and arguments in Python, their differences, types, and how they are used in practice.

#Definition of Parameters
Parameters are the variables listed inside the parentheses in the function definition. They act as placeholders for the values (arguments) that will be passed to the function when it is called. Parameters define what kind of input a function can accept.

#Example:

In [5]:
def greet(name):
    print("Hello, " + name)


#Definition of Arguments
Arguments are the actual values that are passed to the function when it is called. These values are assigned to the corresponding parameters in the function definition.

#Example:

In [6]:
greet("Alice")


Hello, Alice


#3. What are the different ways to define and call a function in Python?
->

Functions are a vital part of any programming language, including Python. They allow us to encapsulate reusable blocks of code that perform specific tasks. In Python, there are multiple ways to define and call functions, offering great flexibility. This assignment will explain the various methods of function definition and function calling in Python with examples and descriptions.

#Defining a Function in Python
A function in Python is defined using the def keyword, followed by the function name, parentheses (), optional parameters, and a colon :. The function body is indented and contains the code that runs when the function is called.

#1. Standard Function Definition
This is the most common way to define a function.
#Example

In [7]:
def greet():
    print("Hello!")


#2. Function with Parameters
You can pass information to functions using parameters.

In [8]:
def greet(name):
    print("Hello,", name)


#3. Function with Return Statement
Functions can return values using the return keyword.

In [9]:
def add(a, b):
    return a + b


#4. Function with Default Parameter Values
You can assign default values to parameters.


In [10]:
def greet(name="Guest"):
    print("Hello,", name)


#4. What is the purpose of the `return` statement in a Python function?
->

In Python, functions are used to perform specific tasks and often need to send the result back to the caller. This is where the return statement becomes essential. The return statement allows a function to produce an output and terminate its execution. In this assignment, we will explore the purpose, usage, and behavior of the return statement in Python functions.

#Definition of return Statement
The return statement in Python is used to exit a function and hand back a value to the function caller. It is typically placed at the end of the function body, though it can appear anywhere inside the function.


#Purpose of the return Statement
1. To Send Output from a Function
The main purpose of return is to provide the result of the function’s processing back to the calling environment.



In [11]:
def add(a, b):
    return a + b

result = add(5, 3)
print(result)  # Output: 8


8


#To Terminate Function Execution
When a return statement is encountered, the function stops executing further instructions.


In [12]:
def demo():
    print("Before return")
    return
    print("After return")  # This line will not execute

demo()


Before return


#To Return Multiple Values
Python allows returning multiple values as a tuple using the return statement.

In [13]:
def get_details():
    name = "Alice"
    age = 25
    return name, age

n, a = get_details()
print(n, a)  # Output: Alice 25


Alice 25


#To Return Different Data Types
The return statement can return any data type including integers, strings, lists, dictionaries, and even other functions.

In [14]:
def get_list():
    return [1, 2, 3, 4]


#Difference Between return and print
return sends the result back to the caller and ends the function.

print only displays the output on the screen and does not affect function logic or return values.

In [15]:
def test():
    print("Hello")
    return "Done"

x = test()  # Prints "Hello", x gets "Done"


Hello


#5. What are iterators in Python and how do they differ from iterables?
->

Python provides powerful tools for managing sequences of data, and two essential concepts in this area are iterables and iterators. While they are closely related, they serve different purposes. Understanding the difference between iterables and iterators is fundamental to working with loops, collections, and efficient data handling in Python. This assignment explains what iterators and iterables are, how they work, and how they differ from each other.

#What is an Iterable?
An iterable is any Python object capable of returning its members one at a time. Examples include lists, tuples, strings, sets, and dictionaries. These objects implement the __iter__() method, which returns an iterator.

#Examples of Iterables:


In [16]:
my_list = [1, 2, 3]
my_string = "Hello"


In [17]:
for char in my_string:
    print(char)


H
e
l
l
o


# What is an Iterator?
An iterator is an object that represents a stream of data; it returns one item at a time when you call the next() function. Iterators implement two special methods:

__iter__() – returns the iterator object itself.

__next__() – returns the next value from the sequence. If there are no more items, it raises a StopIteration exception.

#Creating an Iterator:
You can create an iterator using the iter() function:

In [18]:
my_list = [1, 2, 3]
my_iterator = iter(my_list)

print(next(my_iterator))  # Output: 1
print(next(my_iterator))  # Output: 2


1
2


#Example to Illustrate the Difference.

In [19]:
# Iterable
my_list = [10, 20, 30]

# Convert iterable into an iterator
my_iterator = iter(my_list)

# Accessing elements using iterator
print(next(my_iterator))  # Output: 10
print(next(my_iterator))  # Output: 20
print(next(my_iterator))  # Output: 30


10
20
30


#6. Explain the concept of generators in Python and how they are defined.
->

In Python, generators provide a simple and memory-efficient way to produce a sequence of values. Unlike traditional functions that return a single value and exit, generators yield multiple values one at a time, pausing their execution between each value. They are commonly used in data streaming, large dataset processing, and when working with infinite sequences. This assignment explains the concept of generators, how they work, and how they are defined in Python.


#What is a Generator in Python?
A generator is a type of iterable that generates values on the fly as they are requested, instead of storing all values in memory at once. It is created using a function with the yield keyword instead of return.

Generators maintain internal state, so each time the function is called using next(), it resumes from where it last stopped.

#How to Define a Generator
A generator is defined just like a normal function, but instead of return, it uses the yield keyword to yield values one by one.

#Syntax:

In [20]:
def generator_function():
    yield value


#Example

In [21]:
def count_up_to(n):
    count = 1
    while count <= n:
        yield count
        count += 1


#Calling and Using a Generator
Generators are used just like iterators. You can use the next() function or a for loop.


In [22]:
gen = count_up_to(3)

print(next(gen))  # Output: 1
print(next(gen))  # Output: 2
print(next(gen))  # Output: 3
# print(next(gen))  # This will raise StopIteration


1
2
3


In [23]:
for num in count_up_to(3):
    print(num)


1
2
3


#Advantages of Generators
1. Memory Efficiency: They don’t store all values in memory, which is useful for large datasets.

2. Lazy Evaluation: Values are generated only when needed.

3. Infinite Sequences: They can model infinite data streams.

4. Improved Performance: They often have better performance than using lists or other collections.



#7. What are the advantages of using generators over regular functions?
->

Python provides two primary ways to return data from a function: using regular functions with the return statement, or using generators with the yield keyword. While both approaches are useful, generators offer unique advantages when dealing with large datasets, streams of data, or performance-critical applications. This assignment explains the key advantages of using generators over regular functions in Python.

#What Are Generators?
A generator is a special type of function that uses the yield keyword to produce a sequence of values lazily — one at a time — instead of returning all values at once. When a generator function is called, it returns an iterator object, which can be used to fetch values one by one using next() or a loop.

#1. Memory Efficiency
Generators do not store the entire data set in memory. Instead, they yield one item at a time, which is especially beneficial for large or infinite sequences.

# Example:







In [24]:
def count_up_to(n):
    count = 1
    while count <= n:
        yield count
        count += 1


#2. Lazy Evaluation
Generators use lazy evaluation, meaning values are generated only when requested. This allows the program to start processing data without waiting for the entire sequence to be available.

# Example:

In [25]:
def infinite_numbers():
    num = 1
    while True:
        yield num
        num += 1


#3. Improved Performance
Since generators yield values on demand, they often run faster for large-scale operations or streaming data where intermediate values don’t need to be stored.

#8. What is a lambda function in Python and when is it typically used?
->

In Python, functions are a powerful tool used to organize and reuse code. Alongside regular functions, Python also supports lambda functions, which are small, anonymous functions. Lambda functions are useful for short operations that are used only temporarily, especially when defining a full function would be unnecessarily verbose. In this assignment, we will discuss what a lambda function is, how it is defined, and where it is typically used.

#What is a Lambda Function?
A lambda function in Python is a small anonymous function defined using the lambda keyword. Unlike regular functions that use def and a function name, lambda functions do not require a name (though they can be assigned to a variable). They are limited to a single expression, which is evaluated and returned automatically.

#Syntax of a Lambda Function




In [26]:
lambda arguments: expression


<function __main__.<lambda>(arguments)>

#Key Characteristics of Lambda Functions
* Anonymous: They do not need a name (unless assigned to a variable).

* Single-expression: They can contain only one line of expression.

* Return automatically: No need for the return keyword.

* Used for short, simple tasks.



#Typical Use Cases of Lambda Functions
Lambda functions are commonly used in places where a small function is required temporarily. They are especially useful as arguments to higher-order functions like map(), filter(), and sorted().

# 1. With map() Function
Applies a function to every item in an iterab

In [27]:
numbers = [1, 2, 3, 4]
squared = list(map(lambda x: x * x, numbers))
print(squared)  # Output: [1, 4, 9, 16]


[1, 4, 9, 16]


#2. With filter() Function
Filters items based on a condition.

In [28]:
numbers = [1, 2, 3, 4, 5]
even = list(filter(lambda x: x % 2 == 0, numbers))
print(even)  # Output: [2, 4]


[2, 4]


#3. With sorted() Function
Used to sort based on a custom key.

In [29]:
pairs = [(1, 2), (3, 1), (5, 0)]
sorted_pairs = sorted(pairs, key=lambda x: x[1])
print(sorted_pairs)  # Output: [(5, 0), (3, 1), (1, 2)]


[(5, 0), (3, 1), (1, 2)]


#9. Explain the purpose and usage of the `map()` function in Python.
->

Introduction


Python provides built-in higher-order functions to simplify the processing of iterable data structures. One such powerful and widely used function is the map() function. It allows the application of a specific function to every element of an iterable (like a list or tuple) in an elegant and efficient way. This assignment explores the purpose, syntax, and practical usage of the map() function in Python.

#Purpose of the map() Function
The main purpose of map() is to perform transforma
tions or operations on all elements of an iterable without using explicit loops. It is particularly useful in scenarios where you need to process every element of a collection in the same way.

#Syntax of map()

In [31]:
#map(function, iterable)


#Example of Using map()

In [32]:
def square(x):
    return x * x

numbers = [1, 2, 3, 4]
squared_numbers = map(square, numbers)
print(list(squared_numbers))  # Output: [1, 4, 9, 16]


[1, 4, 9, 16]


#Using map() with Lambda Function
Lambda functions are often used with map() for simple operations.

In [33]:
numbers = [1, 2, 3, 4]
result = map(lambda x: x + 10, numbers)
print(list(result))  # Output: [11, 12, 13, 14]


[11, 12, 13, 14]


#10. What is the difference between `map()`, `reduce()`, and `filter()` functions in Python?

->

Python provides several powerful higher-order functions that allow for functional programming styles, making data transformation and manipulation more concise and readable. Among these functions, the map(), reduce(), and filter() functions are commonly used to perform operations on iterables. Each of these functions serves a different purpose, and understanding their differences and use cases is key to writing clean and efficient Python code. This assignment explores the differences between these three functions.

#1. The map() Function
The map() function applies a specified function to every item in an iterable (such as a list, tuple, etc.) and returns a map object, which is an iterator containing the results.

#Syntax:



In [35]:
#map(function, iterable)


#2. The reduce() Function
The reduce() function, found in the functools module, applies a function cumulatively to the items of an iterable, reducing the iterable to a single value. The function takes two arguments: an accumulator and the current element.

Syntax:

In [37]:
#from functools import reduce
#reduce(function, iterable)


#Example:

In [38]:
from functools import reduce

numbers = [1, 2, 3, 4]
product = reduce(lambda x, y: x * y, numbers)
print(product)  # Output: 24


24


#11. Using pen & Paper write the internal mechanism for sum operation using  reduce function on this given
#list:[47,11,42,13];

->

To calculate the sum of the numbers in the list using the reduce() function.

#Syntax of reduce() Function:

In [40]:
'''from functools import reduce
reduce(function, iterable)'''


'from functools import reduce\nreduce(function, iterable)'

#The reduce() function takes two arguments:

 * function: A binary function (function that takes two arguments).

 * iterable: The collection of items (e.g., a list) to apply the function to.

In this case, we want to sum the elements in the list, so we define a simple lambda function that adds two numbers.

In [41]:
from functools import reduce
numbers = [47, 11, 42, 13]
result = reduce(lambda x, y: x + y, numbers)
print(result)


113


In [42]:
from functools import reduce
numbers = [47, 11, 42, 13]
result = reduce(lambda x, y: x + y, numbers)
print(result)


113


#Step-by-Step Execution of reduce() for Summing
#First Call:

#Initial values:

* x = 47 (first element of the list)

* y = 11 (second element of the list)

Apply the lambda function: 47 + 11 = 58

The result of this call becomes the accumulator for the next step.

#Second Call:

Now, the accumulator is 58 (from the previous step).

* x = 58 (accumulated result from the previous step)

* y = 42 (third element of the list)

Apply the lambda function: 58 + 42 = 100

The result of this call becomes the accumulator for the next step.

#Third Call:

Now, the accumulator is 100 (from the previous step).

* x = 100 (accumulated result from the previous step)

* y = 13 (fourth element of the list)

Apply the lambda function: 100 + 13 = 113

#Final Result:

The final result of the sum operation is 113.

Internal Mechanism Visualized:
Let's visualize this with the following steps:

Step 1: 47 + 11 = 58

Step 2: 58 + 42 = 100

Step 3: 100 + 13 = 113

So, the sum of the list [47, 11, 42, 13] is 113.



#Practical Questions:


#1. Write a Python function that takes a list of numbers as input and returns the sum of all even numbers in the list.

Here's a simple Python function that takes a list of numbers and returns the sum of all even numbers:

In [43]:
def sum_of_evens(numbers):
    return sum(num for num in numbers if num % 2 == 0)


#Example Usage:

In [44]:
my_list = [1, 2, 3, 4, 5, 6]
result = sum_of_evens(my_list)
print(result)  # Output: 12


12


#2. Create a Python function that accepts a string and returns the reverse of that string.


Here's a simple Python function that reverses a given string:


In [45]:
def reverse_string(s):
    return s[::-1]


#Example Usage:

In [46]:
text = "hello"
reversed_text = reverse_string(text)
print(reversed_text)  # Output: "olleh"


olleh


Here's a Python function that takes a list of integers and returns a new list with the square of each number:

In [47]:
def square_numbers(numbers):
    return [num ** 2 for num in numbers]


#Example usage:


In [48]:
original_list = [1, 2, 3, 4, 5]
squared_list = square_numbers(original_list)
print(squared_list)  # Output: [1, 4, 9, 16, 25]


[1, 4, 9, 16, 25]


#4. Write a Python function that checks if a given number is prime or not from 1 to 200

In [49]:
def is_prime(n):
    if n < 2 or n > 200:
        return False  # Not in the range or not prime
    for i in range(2, int(n**0.5) + 1):
        if n % i == 0:
            return False
    return True


#5. Create an iterator class in Python that generates the Fibonacci sequence up to a specified number of terms.

In [50]:
class FibonacciIterator:
    def __init__(self, max_terms):
        self.max_terms = max_terms
        self.count = 0
        self.a, self.b = 0, 1

    def __iter__(self):
        return self

    def __next__(self):
        if self.count >= self.max_terms:
            raise StopIteration
        if self.count == 0:
            self.count += 1
            return 0
        elif self.count == 1:
            self.count += 1
            return 1
        else:
            self.a, self.b = self.b, self.a + self.b
            self.count += 1
            return self.a


#Example Usage:

In [51]:
fib = FibonacciIterator(10)
for num in fib:
    print(num, end=' ')
# Output: 0 1 1 2 3 5 8 13 21 34


0 1 1 1 2 3 5 8 13 21 

#6. Write a generator function in Python that yields the powers of 2 up to a given exponent

In [52]:
def powers_of_two(max_exponent):
    for exp in range(max_exponent + 1):
        yield 2 ** exp


#Example Usage:

In [53]:
for power in powers_of_two(5):
    print(power)


1
2
4
8
16
32


#7. Implement a generator function that reads a file line by line and yields each line as a string.

In [54]:
def read_file_lines(file_path):
    with open(file_path, 'r') as file:
        for line in file:
            yield line.rstrip('\n')  # Remove newline character


#Example usage:


In [56]:
#for line in read_file_lines('example.txt'):
#    print(line)


#8. Use a lambda function in Python to sort a list of tuples based on the second element of each tuple.

In [57]:
# Example list of tuples
tuple_list = [(1, 3), (4, 1), (2, 5), (9, 0)]

# Sort by the second element
sorted_list = sorted(tuple_list, key=lambda x: x[1])

print(sorted_list)


[(9, 0), (4, 1), (1, 3), (2, 5)]


#9. Write a Python program that uses `map()` to convert a list of temperatures from Celsius to Fahrenheit.

In [58]:
# List of temperatures in Celsius
celsius_temps = [0, 20, 30, 37, 100]

# Conversion formula: (C × 9/5) + 32
fahrenheit_temps = list(map(lambda c: (c * 9/5) + 32, celsius_temps))

print(fahrenheit_temps)


[32.0, 68.0, 86.0, 98.6, 212.0]


#10. Create a Python program that uses `filter()` to remove all the vowels from a given string.

In [59]:
# Function to check if a character is a vowel
def is_not_vowel(char):
    return char.lower() not in 'aeiou'

# Given string
input_string = "Hello, how are you?"

# Use filter to remove vowels
filtered_string = ''.join(filter(is_not_vowel, input_string))

print(filtered_string)


Hll, hw r y?


#11

In [60]:
orders = [
    [34587, "Learning Python, Mark Lutz", 4, 40.95],
    [98762, "Programming Python, Mark Lutz", 5, 56.80],
    [77226, "Head First Python, Paul Barry", 3, 32.95],
    [88112, "Einführung in Python3, Bernd Klein", 3, 24.99]
]

result = list(map(
    lambda order: (order[0], order[2]*order[3] + 10 if order[2]*order[3] < 100 else order[2]*order[3]),
    orders
))

print(result)


[(34587, 163.8), (98762, 284.0), (77226, 108.85000000000001), (88112, 84.97)]
